chia/full_node/subscriptions.py

```diff
@@ -1,144 +1,183 @@
 from __future__ import annotations
 
+import contextlib
 import logging
-from dataclasses import dataclass, field
-from typing import Dict, List, Set
+import random
+from dataclasses import dataclass
+from typing import AsyncIterator, List, Set, final
 
 from chia.types.blockchain_format.sized_bytes import bytes32
+from chia.util.db_wrapper import DBWrapper2
 
 log = logging.getLogger(__name__)
 
 
-# The PeerSubscriptions class is essentially a multi-index container. It can be
-# indexed by peer_id, coin_id and puzzle_hash.
-@dataclass(frozen=True)
+@final
+@dataclass
 class PeerSubscriptions:
-    # TODO: use NewType all over to describe these various uses of the same types
-    # Puzzle Hash : Set[Peer ID]
-    _coin_subscriptions: Dict[bytes32, Set[bytes32]] = field(default_factory=dict, init=False)
-    # Puzzle Hash : Set[Peer ID]
-    _ph_subscriptions: Dict[bytes32, Set[bytes32]] = field(default_factory=dict, init=False)
-    # Peer ID: Set[Coin ids]
-    _peer_coin_ids: Dict[bytes32, Set[bytes32]] = field(default_factory=dict, init=False)
-    # Peer ID: Set[puzzle_hash]
-    _peer_puzzle_hash: Dict[bytes32, Set[bytes32]] = field(default_factory=dict, init=False)
-    # Peer ID: subscription count
-    _peer_sub_counter: Dict[bytes32, int] = field(default_factory=dict, init=False)
-
-    def has_ph_subscription(self, ph: bytes32) -> bool:
-        return ph in self._ph_subscriptions
-
-    def has_coin_subscription(self, coin_id: bytes32) -> bool:
-        return coin_id in self._coin_subscriptions
-
-    def add_ph_subscriptions(self, peer_id: bytes32, phs: List[bytes32], max_items: int) -> Set[bytes32]:
-        """
-        returns the puzzle hashes that were actually subscribed to. These may be
-        fewer than requested in case:
-        * there are duplicate puzzle_hashes
-        * some puzzle hashes are already subscribed to
-        * the max_items limit is exceeded
-        """
+    db_wrapper: DBWrapper2
+
+    @classmethod
+    @contextlib.asynccontextmanager
+    async def managed(cls) -> AsyncIterator[PeerSubscriptions]:
+        unique_database_uri = (
+            f"file:db_{cls.__module__}_{cls.__qualname__}_{random.randint(0, 99999999)}?mode=memory&cache=shared"
+        )
+
+        async with DBWrapper2.managed(database=unique_database_uri) as db_wrapper:
```

Comment:
I would highly recommend against using `aiosqlite`, but instead use `sqlite3` directly (the syncronous API). The time you use to do lookups and updates would, most likely, be dominated by the communication with the DB thread.

I would also recommend against using `DBWrapper` (which you wouldn't be able to with the synchronous API anyway).